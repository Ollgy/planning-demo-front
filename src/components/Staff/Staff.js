import React, { PureComponent } from 'react';
import { connect } from 'react-redux';

import { CSSTransition } from 'react-transition-group';
import PersonalForm from '../PersonalForm';
import Input from '../Input';
import AddButton from '../AddButton';
import SectionTitle from '../SectionTitle';
import UserList from '../UserList';
import Popup from '../Popup';
import NewTaskForm from '../NewTaskForm';
import Dashboard from '../Dashboard';
import Cap from '../Cap';
import Preloader from '../Preloader';
import {
  setDialog,
  setCurItem,
  setInitialState,
  inputFilter,
  setViewMode,
  deleteUser,
  getFilterUsers,
  getUsers
} from '../../modules/Staff';
import { D, M, P, V } from '../../const';
import titles from '../../titles.js';
import styles from './Staff.module.css';

class Staff extends PureComponent {
  componentWillUnmount() {
    this.props.setInitialState();
  }

  async componentDidMount() {
    await this.props.getUsers();
  }

  onInputChange = async e => {
    const { getUsers, getFilterUsers } = this.props;
    const { value } = e.target;

    if (value) {
      await getFilterUsers(value);
    } else {
      await getUsers();
    }
  };

  deleteUser = async id => {
    const { deleteUser } = this.props;

    await deleteUser(id);

    this.closePopup();
  };

  openPopup = (msg, curStaffItem) => {
    const { setCurItem, setDialog } = this.props;

    setDialog({
      name: D.popup,
      value: {
        status: true,
        msg: msg || 'popup text'
      }
    });

    setCurItem(curStaffItem);
  };

  closePopup = () => {
    const { setDialog } = this.props;
    const { popup } = this.props.dialog;

    setDialog({
      name: D.popup,
      value: {
        ...popup,
        status: false
      }
    });
  };

  openTaskWindow = (mode, curStaffItem) => {
    const { view, setCurItem, setDialog, setViewMode } = this.props;

    setCurItem(curStaffItem);
    setDialog({
      name: D.newTaskWindow,
      value: {
        status: true
      }
    });

    setViewMode({
      name: view.name,
      mode: mode ? mode : ''
    });
  };

  closeTaskWindow = () => {
    const { view, setDialog, setViewMode } = this.props;

    setDialog({
      name: D.newTaskWindow,
      value: {
        status: false
      }
    });

    setViewMode({
      name: view.name,
      mode: ''
    });
  };

  setStaffMode = (viewName, mode, curStaffItem) => {
    const { setCurItem, setViewMode } = this.props;

    setCurItem(curStaffItem);
    setViewMode({
      name: viewName,
      mode: mode ? mode : ''
    });
  };

  returnToList = async e => {
    e.preventDefault();

    const { getUsers, setViewMode } = this.props;

    await getUsers();

    setViewMode({
      name: V.userlist,
      mode: M.read
    });
  };

  onClickOk = () => {
    const setDialog = this.props;
    const { popup } = this.props.dialog;

    setDialog({
      name: D.popup,
      value: {
        ...popup.value,
        userAnswer: true
      }
    });
  };

  render() {
    const { userPermission } = this.props;
    const {
      curStaffItem,
      staff,
      filterValue,
      dialog,
      view,
      personalFormMode,
      isLoading
    } = this.props;
    const { popup, newTaskWindow } = dialog;

    return !isLoading ? (
      <section className={styles.sectionContainer}>
        <CSSTransition
          in={popup.status}
          timeout={300}
          mountOnEnter={true}
          unmountOnExit={true}
          classNames={{
            appear: styles.appear,
            enter: styles.enter,
            enterActive: styles.enterActive,
            exit: styles.exit,
            exitActive: styles.exitActive
          }}
        >
          <Popup
            msg={popup.msg}
            closeClick={this.closePopup}
            okClick={this.deleteUser.bind(null, curStaffItem.id)}
          />
        </CSSTransition>
        <CSSTransition
          in={newTaskWindow.status}
          timeout={300}
          mountOnEnter={true}
          unmountOnExit={true}
          classNames={{
            appear: styles.appear,
            enter: styles.enter,
            enterActive: styles.enterActive,
            exit: styles.exit,
            exitActive: styles.exitActive
          }}
        >
          <NewTaskForm
            mode={view.mode}
            taskData={{ executorId: curStaffItem.id }}
            onClose={this.closeTaskWindow}
          />
        </CSSTransition>
        <SectionTitle children="Сотрудники" size="m" />
        <header className={styles.header}>
          {view.name !== V.userlist ? (
            <div className={styles.returnBlock}>
              <a
                className={styles.returnLink}
                onClick={this.returnToList}
                href=" "
                children="Вернуться к списку"
              />
            </div>
          ) : null}
          <div
            className={`${styles.searchBlock} ${
              view.name !== V.userlist ? styles.disabled : null
            }`}
          >
            <Input
              type="text"
              name="search"
              value={filterValue}
              readonly={false}
              placeholder="Поиск"
              onChange={this.onInputChange}
            />
          </div>
          <div
            className={`${styles.addButtonBlock} ${
              view.name !== V.userlist ? styles.disabled : null
            }`}
            data-tooltip="Добавить&#10;нового&#10;сотрудника"
          >
            <AddButton
              onClick={this.setStaffMode.bind(null, V.profile, M.create)}
            />
          </div>
        </header>
        {view.name === V.profile ? (
          <div className={styles.personalFormWrap}>
            <SectionTitle children={titles[personalFormMode]} size="s" />
            <PersonalForm
              permission={P.manager}
              startMode={view.mode}
              user={curStaffItem}
            />
          </div>
        ) : view.name === V.userlist ? (
          staff.length ? (
            <UserList
              list={staff}
              activities={{
                createUserTask: this.openTaskWindow.bind(null, M.create),
                deleteUser: this.deleteUser,
                openPopup: this.openPopup,
                closePopup: this.closePopup,
                updateTaskWindow: this.openTaskWindow.bind(null, M.update),
                updateUser: this.setStaffMode.bind(null, V.profile, M.update),
                getTaskList: this.setStaffMode.bind(null, V.tasklist, null)
              }}
            />
          ) : (
            <Cap text="Сотрудников не найдено" />
          )
        ) : (
          <div className={styles.dashboardWrap}>
            <Dashboard
              userId={curStaffItem.id}
              userName={`${curStaffItem.firstName} ${curStaffItem.lastName}`}
              userPermission={userPermission}
            />
          </div>
        )}
      </section>
    ) : (
      <Preloader />
    );
  }
}

const mapStateToProps = state => ({
  filterValue: state.staff.datalist.filterValue,
  curStaffItem: state.staff.datalist.curStaffItem,
  staff: state.staff.datalist.staff,
  dialog: state.staff.dialog.data,
  view: state.staff.view.data,
  isLoading: state.staff.getUsers.isLoading,
  personalFormMode: state.personalForm.view.mode
});

const mapDispatchToProps = dispatch => ({
  getUsers: () => dispatch(getUsers()),
  getFilterUsers: arg => dispatch(getFilterUsers(arg)),
  deleteUser: arg => dispatch(deleteUser(arg)),
  setViewMode: arg => dispatch(setViewMode(arg)),
  setDialog: arg => dispatch(setDialog(arg)),
  inputFilter: arg => dispatch(inputFilter(arg)),
  setCurItem: arg => dispatch(setCurItem(arg)),
  setInitialState: () => dispatch(setInitialState())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Staff);
