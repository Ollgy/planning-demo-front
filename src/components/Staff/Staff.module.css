.sectionContainer {
  height: 100%;
  padding-top: 20px;
  overflow-y: auto;
}

.header {
  display: flex;
  width: 100%;
  justify-content: flex-end;
  align-items: center;
  padding: 10px 30px;
  background-color: #e6e6e6;
}

.addButtonBlock {
  position: relative;
}

.searchBlock,
.returnBlock {
  position: relative;
  margin-right: 20px;
}

.disabled:after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 1;
  background-color: #e6e6e6;
  opacity: 0.8;
}

.returnLink {
  padding: 10px;
  color: #707070;
}

.personalFormWrap {
  padding-top: 20px;
}

.dashboardWrap {
  min-height: 200px;
}

.enter {
  opacity: 0.01;
}

.enter.enterActive {
  opacity: 1;
  transition: opacity 300ms ease-in;
}

.exit {
  opacity: 1;
}

.exit.exitActive {
  opacity: 0.01;
  transition: opacity 300ms ease-in;
}
