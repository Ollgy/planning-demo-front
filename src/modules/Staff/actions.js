import { createAction } from 'redux-actions';
import api from '../../api';

export const setCurItem = createAction('SET_CUR_ITEM_STAFF');
export const setDialog = createAction('SET_DIALOG_STAFF');
export const setViewMode = createAction('SET_VIEW_MODE_STAFF');
export const setInitialState = createAction('SET_INITIAL_STATE_STAFF');
export const inputFilter = createAction('INPUT_FORM_FILTER_STAFF');

export const getFilterUsersRequest = createAction('GET_FILTER_USERS_REQUEST');
export const getFilterUsersSuccess = createAction('GET_FILTER_USERS_SUCCESS');
export const getFilterUsersFailure = createAction('GET_FILTER_USERS_FAILURE');

export const getFilterUsers = str => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(getFilterUsersRequest());

      api
        .getFilterUsers(str)
        .then(users => {
          dispatch(getFilterUsersSuccess(users));
          resolve();
        })
        .catch(error => {
          dispatch(getFilterUsersFailure(error));
          reject();
        });
    });
};

export const deleteUserRequest = createAction('DELETE_USER_REQUEST');
export const deleteUserSuccess = createAction('DELETE_USER_SUCCESS');
export const deleteUserFailure = createAction('DELETE_USER_FAILURE');

export const deleteUser = userId => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(deleteUserRequest());

      api
        .deleteUser(userId)
        .then(users => {
          dispatch(deleteUserSuccess(users));
          resolve();
        })
        .catch(error => {
          dispatch(deleteUserFailure(error));
          reject();
        });
    });
};

export const getUsersRequest = createAction('GET_USERS_REQUEST');
export const getUsersSuccess = createAction('GET_USERS_SUCCESS');
export const getUsersFailure = createAction('GET_USERS_FAILURE');

export const getUsers = () => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(getUsersRequest());

      api
        .getUsers()
        .then(users => {
          dispatch(getUsersSuccess(users));
          resolve();
        })
        .catch(error => {
          dispatch(getUsersFailure(error));
          reject();
        });
    });
};
