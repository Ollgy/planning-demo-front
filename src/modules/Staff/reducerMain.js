import { combineReducers } from 'redux';
import delUser from './reducers/delUser';
import filterUsers from './reducers/filterUsers';
import getUsers from './reducers/getUsers';
import datalist from './reducers/datalist';
import dialog from './reducers/dialog';
import view from './reducers/view';

export default combineReducers({
  datalist,
  dialog,
  view,
  delUser,
  filterUsers,
  getUsers
});
