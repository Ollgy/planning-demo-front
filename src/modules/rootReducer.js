import { combineReducers } from 'redux';
import authFromToken from './Auth';
import login from './Login';
import loginForm from './LoginForm';
import profile from './Profile';
import personalForm from './PersonalForm';
import changePswrdForm from './ChangePswrdForm';
import newTaskForm from './NewTaskForm';
import notes from './Notes';
import dashboard from './Dashboard';
import staff from './Staff';
import cleanBase from './CleanBase';

export default combineReducers({
  authFromToken,
  login,
  loginForm,
  profile,
  personalForm,
  changePswrdForm,
  newTaskForm,
  notes,
  dashboard,
  staff,
  cleanBase
});
