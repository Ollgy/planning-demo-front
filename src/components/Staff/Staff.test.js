import React from 'react';
import Staff from './Staff';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { createStore, compose, applyMiddleware } from 'redux';
import rootReducer from '../../modules/rootReducer';
import thunk from 'redux-thunk';

const store = createStore(rootReducer, {}, compose(applyMiddleware(thunk)));

describe('<Staff/> UI component', () => {
  const shallowExpect = compose(
    expect,
    toJSON,
    shallow
  );

  it('Renders correctly', () => {
    shallowExpect(<Staff store={store} />).toMatchSnapshot();
  });

  const wrapper = shallow(<Staff store={store} />);
  const inner = wrapper.children().first();

  it('contains staff in props', () => {
    expect(inner.props().staff).toBeTruthy();
  });
});
