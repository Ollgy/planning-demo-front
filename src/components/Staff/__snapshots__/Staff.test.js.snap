// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Staff/> UI component Renders correctly 1`] = `
<ContextProvider
  value={null}
>
  <Staff
    curStaffItem={Object {}}
    deleteUser={[Function]}
    dialog={
      Object {
        "newTaskWindow": Object {
          "status": false,
        },
        "popup": Object {
          "msg": "",
          "status": false,
          "userAnswer": false,
        },
      }
    }
    getFilterUsers={[Function]}
    getUsers={[Function]}
    inputFilter={[Function]}
    isLoading={false}
    personalFormMode="read"
    setCurItem={[Function]}
    setDialog={[Function]}
    setInitialState={[Function]}
    setViewMode={[Function]}
    staff={Array []}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    view={
      Object {
        "mode": "read",
        "name": "userlist",
      }
    }
  />
</ContextProvider>
`;
